keps/sig-api-machinery/2309-idl/README.md

```diff
@@ -0,0 +1,609 @@
+  compile existing pseudo-IDL to CKDL, and then reverse the CKDL back to
+  KDL, which means by-and-large the migration can be automated, with
+  manual intervention solely to supplement existing type information or
+  for aesthetic purposes.
+
+* **Dev tooling**: unlike the existing pseudo-IDL, there will be no
+  existing syntax highlighting/jump-to-definition/etc out of the box.
+
+  * Basic syntax highlighting for several popular editors (VIM, VS
+    Code/other textmate-grammar-based editors) is fairly achievable
+    short-term (the [repository][repo] contains an implementation of VIM
+    syntax highlighting).
+
+  * Supplemental functionality via a language server implementation could
+    restore the rest of this functionality in most popular editors, but
+    this would involve significantly more work, and may not be worth it in
+    the short term.
+
+### Compatibility
+
+As the output format (CKDL) is serialized as proto blobs, proper
+discipline with changes to CDKL should mean we get the built-in
+compatibility guarnatees of protobuf.  This means that CKDL files compiled
+with older compilers should work automatically with newer tooling, and
+vice versa.
+
+Likewise, since marker definitions are imported, it's possible to add new
+markers to the language without changing the syntax or compiler.
+Similarly, since marker defintions are effectively syntactic sugar over
+proto message definitions, the markers *themselves* benefit from the same
+guarantees as CKDL itself around compatibility.
+
+This leaves the KDL syntax itself.  It's expected that the syntax itself
+will not change frequently.  In the case of major syntax changes, the
+CKDL-to-KDL converter may be used to migrate source files.
+
+*If* major syntax changes are needed, a second version of the language may
+be introduced, with an equivalent major version in the file extensions
+(e.g. `file.kdl2`).
+
+## Design Details
+
+See the [directxman12/idl][repo] repository for a working implementation
+of the compiler and generators, as well as additional documentation.
+
+[repo]: https://github.com/directxman12/idl
+
+### Test Plan
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+### Graduation Criteria
+
+## Implementation History
+
+## Alternatives
```

Comment:
Since the proposal assumes an IDL is necessary, I expected alternatives to include options that don’t include idls, like tools that check go types for conventions more effectively.

Reply:
I'm happy to add a bit more of an explanation here, but that's what we do *today* (it's what controller-tools does, for instance), and I figured the reasons were largely covered by the "What problems do we have today" section.  Is there something there that's unclear?

Reply:
You have described all the problems above, but the solution is phrased as "only an IDL can solve the problem", vs "additional constraints on the tooling we have around go types".  I expected that to be addressed here as "we considered continuing to improve the existing tooling but X doesn't work because Y"

For instance, up above:

> Different parts of the pseudo-IDL are split across different files (types vs validation vs defaulting), making it more difficult to review holistically.

But there is a proposal I've seen to move defaulting to the types themselves.  Won't completely solve the problem, but addresses this issue.  I think because that issue is smaller, it's likely to see some movement (vs this much larger effort).  To make the case for the IDL, it is most effective if you can show that the IDL solves those problems and all their complexity more efficiently (i.e. your IDL is going to solve defaulting, so we can spend X doing it with a one off vs Y as part of a larger effort)

Reply:
> But there is a proposal I've seen to move defaulting to the types themselves

I'm well aware of that proposal -- it's pulling a feature that I wrote for controller-tools to k/k (i.e. it's something that we already do today for CRDs).

> Won't completely solve the problem, but addresses this issue. I think because that issue is smaller, it's likely to see some movement (vs this much larger effort).

It solves one part of the issue, while ignoring the rest.  Fundamentally, "we don't have declarative defaulting/validation" is just *one* part of the problem.  A *very* significant other part of the problem is "the devex of writing CRDs is terrible", and in this, our "magic comments / 4 different ways of specifying the same thing" is fundamentally limited.

Having spent the last 2+ years trying to improve the devex of api machinery from the "outside", it's not at all clear to me how more tooling on the existing system would fix those fundamental issues.

If we *just* solve this for k/k piecemeal, we end up ignoring the end users.  If we *just* solve this in controller-tools, we improve the overall devex in some ways, but we exacerbate the tooling drift, making it worse in others.  That's why, IMO, we need to solve these two pieces together (FWIW, I've put that last bit in the KEP: https://github.com/DirectXMan12/enhancements/blob/kep/idl/keps/sig-api-machinery/2309-idl/README.md#addendum-why-do-we-need-to-do-this-in-kk)

Reply:
I like @DirectXMan12's argument here.

Reply:
A big part of the problem, I think, is that the pseudo-IDL files serve 2 masters.  They are both the IDL and the compileable Go implementation of that IDL.

Thought experiment:

Assume we took all of the types.go files and renamed them to types.pidl (pseudo IDL).  Now they are not part of the literal build any more.  We could then achieve some (how much?) of the goals of this KEP.  We could remove the struct tags (duplicate names).  We could add more markers to indicate the various facts we want to assert about each type.  We could limit the Go features we allow - e.g. no using pointers, we will infer pointer-ness when we generate "real" Go code.  We could use Go's tooling for parsing and processing syntax (since it is still a subset of Go) and validate ourselves that the subset is adhered to.  At the end we emit types.go files which look an awful lot like today's files, but now humans don't edit them.

It's not very pretty, for sure.  Some problems persist (e.g. breaking comment blocks) and if we do enough custom validation, it effectively becomes a bespoke interpreter.  But we *could* do it, right?  We could shoehorn declarative defaults and validation.  We could even compile to CKDL and lean on the rest of your proposal.

So the question is: Why is this less-good than KDL as a net-new grammar?

Reply:
> A big part of the problem, I think, is that the pseudo-IDL files serve 2 purposes. They are both the IDL and the compileable Go implementation of that IDL.

Yeah, this is a big factor -- it's one of the things I tried to talk about when I mentioned "wire format vs representation" (there used to be a bigger section on that, but it got toned down a bit to avoid confusion over "are we gonna replace the Go structs with different ones")

> So the question is: Why is this less-good than KDL as a net-new grammar?

I've grappled with this a bit myself, and here's what I managed to convince myself of:

1. Markers are janky (I've talked a lot about this already -- this is a big point for me)
    * They're easy to screw up, hard to properly validate, and require some real sketch parsing code to actually extract from files.
    * They look like they're bolted on, which is suboptimal for teaching, especially for things that should be core constructs, IMO (kinds, list-maps, enums, unions)

2. The language is fundamentally missing constructs (e.g. enums, kinds, etc).  Some of these, like kinds, are possible to retrofit using markers, but it still feels clunky to me.  Others, though, are not so easy.  Take enums.  Go doesn't have a concept of enums -- the closest you can come is constants with iota.  So now, you're trying to parse and then look for all constants that are of a particular type.  This requires a bit more type-checking, and is not particularly obvious of an idiom, *especially* for folks coming from languages that aren't Go or C.

3. "It's like Go, but not" is perhaps a useful way to teach people who already know Go and k8s, but I'd argue that it's not useful to folks who don't know Go & k8s -- saying "just go learn Go" won't work, because some Go stuff is disallowed, especially stuff that are common idioms (floats, pointers, etc).

<details>
<summary>the uncanny valley half-argument</summary>

4 -- "It's like Go, but not" causes frustration.  I've saved this argument for last, because it holds a bit less water, but languages that billed as "X but not" garner an uncanny valley effect -- you try to do stuff you're used to, and get frustrated when it doesn't work.  I've encountered this while supporting pseudo-IDL users ("wait, why can't I use an interface here?", "wait, it's just Go code, why can't I import & embed these random types from another package"), as well as myself to some extent with other languages (skylark vs python, elm vs haskell).  By itself, the argument doesn't hold a ton of water, but it's worth considering along with the other arguments IMO.

</details>

Fundamentally, I think we lose out here on readability (positioning of "important" vs "supplementary" information), teachability, and to some extent being able to give better error messages (with the parser, you get errors that are phrased in Go terms, not in k8s terms).

If this were *just* about k/k, then maybe that'd be easier (I'm still not convinced, but this *might* hold some water), but factoring in non-k/k uses, and *especially* non-Go uses, I think the IDL is a lot better in the long run.

Reply:
Also, FWIW, I think people are *vastly* underestimating the amount of custom parsing code that goes into our current system that lays on top of the Go parser (not denying that we get *some* nice stuff with the Go parser, but it's not as straightforward as it seems):

(and yes, pre-emptively, LoC doesn't tell the whole story, but it's a decent proxy here):

Excluding generated help files & tests (there's a lot of those, but the equivalent ckdl tests don't exist yet), but including the CRD generator because it holds a lot of the k8s semantic type-checking logic and such:

```
sollyross@G💻 ~[go:sigs.k8s.io/controller-gen] $  tokei --type Go -e testdata -e zz_generated.markerhelp.go -e '*_test.go' pkg/{genall,loader,markers,crd} cmd/{controller-gen,helpgen}
===============================================================================
 Language            Files        Lines         Code     Comments       Blanks
===============================================================================
 Go                     44         8098         5075         2118          905
===============================================================================
```

Plus we also maintain (note that gengo punts on stuff like marker parsing (forcing it into individual generators instead) and performance optimizations, so these numbers are actually a bit higher):

```
sollyross@G💻 ~[go:gengo] $ tokei --sort Code --type Go -e testdata -e '*_test.go' -e examples
===============================================================================
 Language            Files        Lines         Code     Comments       Blanks
===============================================================================
 Go                     21         3597         2342          943          312
===============================================================================
```

(plus at hundreds of lines of Makefile for performance reasons)

versus (in progress, but contains largely the base functionality):

```
sollyross@G💻 ~[go:idl] $ tokei --type Go kdlc
===============================================================================
 Language            Files        Lines         Code     Comments       Blanks
===============================================================================
 Go                     15         4306         3793          186          327
===============================================================================
```

Reply:
None of this is to say that the IDL is necessarily gonna be simpler code-wise, but the stuff we have now is of non-trivial complexity, adding more validation on probably isn't gonna lessen that.

Reply:
Plus, if we *ever* decide we want to change the pidl, even a little bit, we suddenly then own a fork of the entire Go parser, lexer, gofmt, etc.